osutil/sys/runas.go

```diff
@@ -0,0 +1,90 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package sys
+
+import (
+	"fmt"
+	"runtime"
+	"syscall"
+)
+
+// UnrecoverableError is an error that flags that things have Gone Wrong, the
+// runtime is in a bad state, and you should really quit. The intention is that
+// if you're trying to recover from a panic and find that the value of the panic
+// is an UnrecoverableError, you should just exit ASAP.
+type UnrecoverableError struct {
+	Call string
+	Err  error
+}
+
+func (e UnrecoverableError) Error() string {
+	return fmt.Sprintf("%s: %v", e.Call, e.Err)
+}
+
+// RunAsUidGid starts a goroutine, pins it to the OS thread, sets euid and egid,
+// and runs the function; after the function returns, it restores euid and egid.
+//
+// If restoring the original euid and egid fails this function will panic with
+// an UnrecoverableError, and you should _not_ try to recover from it: the
+// runtime itself is going to be in trouble.
+func RunAsUidGid(uid UserID, gid GroupID, f func() error) error {
+	ch := make(chan error, 1)
+	go func() {
+		// from the docs:
+		//   until the goroutine exits or calls UnlockOSThread, it will
+		//   always execute in this thread, and no other goroutine can.
+		// that last bit means it's safe to setuid/setgid in here, as no
+		// other code will run.
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+
+		ruid := Getuid()
+		rgid := Getgid()
+
+		// change GID
+		if _, _, errno := syscall.RawSyscall(_SYS_SETREGID, FlagID, uintptr(gid), 0); errno != 0 {
```

Comment:
It seems later go versions support syscall.Setregid() and friedns so hopefully we don't need the RawSyscall? I got a bit overboard (sorry!) and wonder if this might make sense?
```go
diff --git a/osutil/sys/runas.go b/osutil/sys/runas.go
index 712f83c589..4f13884c8c 100644
--- a/osutil/sys/runas.go
+++ b/osutil/sys/runas.go
@@ -25,17 +25,17 @@ import (
        "syscall"
 )
 
-// UnrecoverableError is an error that flags that things have Gone Wrong, the
-// runtime is in a bad state, and you should really quit. The intention is that
-// if you're trying to recover from a panic and find that the value of the panic
-// is an UnrecoverableError, you should just exit ASAP.
-type UnrecoverableError struct {
-       Call string
-       Err  error
-}
-
-func (e UnrecoverableError) Error() string {
-       return fmt.Sprintf("%s: %v", e.Call, e.Err)
+func composeErr(prefix1 string, err1 error, prefix2 string, err2 error) error {
+       switch {
+       case err1 != nil && err2 != nil:
+               return fmt.Errorf("%v: %v and %v: %v", prefix1, err1, prefix2, err2)
+       case err1 != nil:
+               return fmt.Errorf("%v: %v", prefix1, err1)
+       case err2 != nil:
+               return fmt.Errorf("%v: %v", prefix2, err2)
+       default:
+               return nil
+       }
 }
 
 // RunAsUidGid starts a goroutine, pins it to the OS thread, sets euid and egid,
@@ -59,32 +59,30 @@ func RunAsUidGid(uid UserID, gid GroupID, f func() error) error {
                rgid := Getgid()
 
                // change GID
-               if _, _, errno := syscall.RawSyscall(_SYS_SETREGID, FlagID, uintptr(gid), 0); errno != 0 {
-                       ch <- fmt.Errorf("setregid: %v", errno)
+               if err := syscall.Setregid(int(gid), 0); err != nil {
+                       ch <- fmt.Errorf("cannot set setregid: %v", err)
                        return
                }
 
-               // make sure we restore GID again
-               defer func() {
-                       if _, _, errno := syscall.RawSyscall(_SYS_SETREGID, FlagID, uintptr(rgid), 0); errno != 0 {
-                               panic(UnrecoverableError{Call: "setregid", Err: errno})
-                       }
-               }()
-
                // change UID
-               if _, _, errno := syscall.RawSyscall(_SYS_SETREUID, FlagID, uintptr(uid), 0); errno != 0 {
-                       ch <- fmt.Errorf("setreuid: %v", errno)
+               if err := syscall.Setreuid(int(uid), 0); err != nil {
+                       ch <- fmt.Errorf("cannot setreuid: %v", err)
                        return
                }
 
egon@top:~/devel/go/src/github.com/snapcore/snapd$ git --no-pager diff
diff --git a/osutil/sys/runas.go b/osutil/sys/runas.go
index 712f83c589..4f13884c8c 100644
--- a/osutil/sys/runas.go
+++ b/osutil/sys/runas.go
@@ -25,17 +25,17 @@ import (
 	"syscall"
 )
 
-// UnrecoverableError is an error that flags that things have Gone Wrong, the
-// runtime is in a bad state, and you should really quit. The intention is that
-// if you're trying to recover from a panic and find that the value of the panic
-// is an UnrecoverableError, you should just exit ASAP.
-type UnrecoverableError struct {
-	Call string
-	Err  error
-}
-
-func (e UnrecoverableError) Error() string {
-	return fmt.Sprintf("%s: %v", e.Call, e.Err)
+func composeErr(prefix1 string, err1 error, prefix2 string, err2 error) error {
+	switch {
+	case err1 != nil && err2 != nil:
+		return fmt.Errorf("%v: %v and %v: %v", prefix1, err1, prefix2, err2)
+	case err1 != nil:
+		return fmt.Errorf("%v: %v", prefix1, err1)
+	case err2 != nil:
+		return fmt.Errorf("%v: %v", prefix2, err2)
+	default:
+		return nil
+	}
 }
 
 // RunAsUidGid starts a goroutine, pins it to the OS thread, sets euid and egid,
@@ -59,32 +59,30 @@ func RunAsUidGid(uid UserID, gid GroupID, f func() error) error {
 		rgid := Getgid()
 
 		// change GID
-		if _, _, errno := syscall.RawSyscall(_SYS_SETREGID, FlagID, uintptr(gid), 0); errno != 0 {
-			ch <- fmt.Errorf("setregid: %v", errno)
+		if err := syscall.Setregid(int(gid), 0); err != nil {
+			ch <- fmt.Errorf("cannot set setregid: %v", err)
 			return
 		}
 
-		// make sure we restore GID again
-		defer func() {
-			if _, _, errno := syscall.RawSyscall(_SYS_SETREGID, FlagID, uintptr(rgid), 0); errno != 0 {
-				panic(UnrecoverableError{Call: "setregid", Err: errno})
-			}
-		}()
-
 		// change UID
-		if _, _, errno := syscall.RawSyscall(_SYS_SETREUID, FlagID, uintptr(uid), 0); errno != 0 {
-			ch <- fmt.Errorf("setreuid: %v", errno)
+		if err := syscall.Setreuid(int(uid), 0); err != nil {
+			ch <- fmt.Errorf("cannot setreuid: %v", err)
 			return
 		}
 
+		funcErr := f()
+
+		// make sure we restore GID again
+		if err := syscall.Setregid(int(rgid), 0); err != nil {
+			ch <- composeErr("cannot run func", funcErr, "cannot restore regid", err)
+		}
+
 		// make sure we restore UID again
-		defer func() {
-			if _, _, errno := syscall.RawSyscall(_SYS_SETREUID, FlagID, uintptr(ruid), 0); errno != 0 {
-				panic(UnrecoverableError{Call: "setreuid", Err: errno})
-			}
-		}()
+		if err := syscall.Setreuid(int(ruid), 0); err != nil {
+			ch <- composeErr("cannot run func", funcErr, "cannot restore reuid", err)
+		}
 
-		ch <- f()
+		ch <- funcErr
 	}()
 	return <-ch
 }
diff --git a/osutil/sys/sysnum_16_linux.go b/osutil/sys/sysnum_16_linux.go
index dd425854b6..6d01a57abf 100644
--- a/osutil/sys/sysnum_16_linux.go
+++ b/osutil/sys/sysnum_16_linux.go
@@ -27,10 +27,8 @@ import "syscall"
 // (and so the 32 bit version is called getuid32 etc)
 
 const (
-	_SYS_GETUID   = syscall.SYS_GETUID32
-	_SYS_GETGID   = syscall.SYS_GETGID32
-	_SYS_GETEUID  = syscall.SYS_GETEUID32
-	_SYS_GETEGID  = syscall.SYS_GETEGID32
-	_SYS_SETREUID = syscall.SYS_SETREUID32
-	_SYS_SETREGID = syscall.SYS_SETREGID32
+	_SYS_GETUID  = syscall.SYS_GETUID32
+	_SYS_GETGID  = syscall.SYS_GETGID32
+	_SYS_GETEUID = syscall.SYS_GETEUID32
+	_SYS_GETEGID = syscall.SYS_GETEGID32
 )
diff --git a/osutil/sys/sysnum_32_linux.go b/osutil/sys/sysnum_32_linux.go
index f2258e6678..36f67fc48a 100644
--- a/osutil/sys/sysnum_32_linux.go
+++ b/osutil/sys/sysnum_32_linux.go
@@ -26,10 +26,8 @@ import "syscall"
 // these are the constants for where getuid et al are already 32-bit
 
 const (
-	_SYS_GETUID   = syscall.SYS_GETUID
-	_SYS_GETGID   = syscall.SYS_GETGID
-	_SYS_GETEUID  = syscall.SYS_GETEUID
-	_SYS_GETEGID  = syscall.SYS_GETEGID
-	_SYS_SETREUID = syscall.SYS_SETREUID
-	_SYS_SETREGID = syscall.SYS_SETREGID
+	_SYS_GETUID  = syscall.SYS_GETUID
+	_SYS_GETGID  = syscall.SYS_GETGID
+	_SYS_GETEUID = syscall.SYS_GETEUID
+	_SYS_GETEGID = syscall.SYS_GETEGID
 )
diff --git a/osutil/sys/sysnum_darwin.go b/osutil/sys/sysnum_darwin.go
index 8383dd6e40..14cbc986f3 100644
--- a/osutil/sys/sysnum_darwin.go
+++ b/osutil/sys/sysnum_darwin.go
@@ -27,6 +27,4 @@ const (
 	_SYS_GETEUID  = unix.SYS_GETEUID
 	_SYS_GETEGID  = unix.SYS_GETEGID
 	_SYS_FCHOWNAT = unix.SYS_FCHOWNAT
-	_SYS_SETREUID = unix.SYS_SETREUID
-	_SYS_SETREGID = unix.SYS_SETREGID
 )
```
(also attached)
[tweak-runas.go.gz](https://github.com/snapcore/snapd/files/12039216/tweak-runas.go.gz)

Reply:
Except that in my local qemu run the test fails with that :(